# Logit regressions

In [1]:
from sklearn import linear_model
from sklearn import tree
from sklearn import ensemble

from sklearn.datasets import make_blobs
from sklearn.model_selection import train_test_split
from sklearn import metrics
from sklearn.inspection import permutation_importance
from sklearn.preprocessing import StandardScaler

# common visualization module
from matplotlib import pyplot as plt
import plotly.express as px
import seaborn as sns
sns.set()

# numeric library
import numpy as np

import os
import pandas as pd
from time import time as timer
import tarfile

import tensorflow as tf

%matplotlib inline
from matplotlib import animation
from IPython.display import HTML

In [ ]:
file_path = 'C:/Users/Gaëlle/Documents/_CAS applied data science/3. Module 3 Data analysis and machine learning/Project/Models/final_GMA.csv'
df = pd.read_csv(file_path)

In [ ]:
column_names=df.columns.tolist()
column_names

['c_section_rate',
 'I.1.16.M',
 'A.5.1.F',
 'A.7.14.M',
 'L.6.4.F',
 'KostAWLFA',
 'L.5.5.F',
 'A.7.1.F',
 'D.2.6.P',
 'J.1.4.M',
 'L.6.8.F',
 'J.1.5.M',
 'A.7.4.F',
 'pPatLKP',
 'KostKVGStatA',
 'AnlKVGStatA',
 'A.7.7.M',
 'F.3.7.M',
 'D.1.2.M',
 'B.1.9.M',
 'H.1.5.P',
 'I.1.8.M',
 'L.5.3.F',
 'J.3.1.P',
 'H.4.2.P',
 'A.7.15.M',
 'A.7.18.P',
 'B.1.14.P',
 'E.4.6.M',
 'E.4.16.P',
 'H.1.9.P',
 'D.1.1.M',
 'KostOKPAmbA',
 'G.5.3.P',
 'A.7.24.F',
 'Gebs',
 'A.7.28.M',
 'AustStatMSA',
 'F.3.3.M',
 'A.3.2.V',
 'F.1.4.F',
 'A.7.13.M',
 'I.1.9.M',
 'G.2.2.F',
 'H.1.7.M',
 'E.4.2.M',
 'StdBelA',
 'E.5.4.F',
 'H.3.2.F',
 'E.3.1.F',
 'A.1.14.P',
 'L.6.1.F',
 'E.4.15.M',
 'A.7.11.V',
 'I.2.7.M',
 'KostAmbA',
 'F.3.9.F',
 'A.7.8.M',
 'A.7.12.V',
 'J.1.1.M',
 'G.3.5.P',
 'H.3.3.F',
 'F.1.5.X',
 'C.1.5.S',
 'F.2.8.M',
 'G.6.3.F',
 'E.4.11.F',
 'B.1.11.M',
 'A.7.29.M',
 'E.4.1.V',
 'A.7.6.F',
 'J.1.3.P',
 'A.7.27.M',
 'G.6.4.F',
 'AustStatMST',
 'I.1.18.M',
 'D.4.5.M',
 'G.4.4.M',
 'F.2.10.M',
 'I.1

## Use dataset with the c-section threshold

In [ ]:
# Let's standardize all the variables
scaler = StandardScaler()
df_stand_array = scaler.fit_transform(df)
# back to a DataFrame
df = pd.DataFrame(df_stand_array, columns=df.columns)

## Creation of the X, Y, train and test

In [ ]:
# Creation of the dependant variable
target_field = 'threshold'
Y = df[target_field].copy()

In [ ]:
# Set of independant variables
X = df.drop(columns=['threshold','c_section_rate'])

In [ ]:
X.shape

(298, 466)

In [ ]:
x = X.to_numpy().astype(np.float32)
y = Y.to_numpy().astype(np.float32)

In [ ]:
# train and test sets
x_train, x_test, y_train, y_test = train_test_split(x, y, test_size=0.2)

### Fit the model

In [ ]:
logistic = linear_model.LogisticRegression(solver='sag', max_iter=200,
                             multi_class='ovr',)
logistic.fit(x_train, y_train)

ValueError: Unknown label type: continuous. Maybe you are trying to fit a classifier, which expects discrete classes on a regression target with continuous values.

In [ ]:
# Evaluate MSE, MAE, and R2 on train and test datasets
yy_test = logistic.predict(x_test)
yy_train = logistic.predict(x_train)

print('train R2 =', logistic.score(x_train, y_train))
print('test R2 =', logistic.score(x_test, y_test))


In [ ]:
# Coefficient matrix
cm=metrics.confusion_matrix( y_test, yy_test)
cm

In [ ]:
# Confusion Matrix Evaluation Metrics
from sklearn.metrics import classification_report
target_names = ['Below threshold', 'Above threshold']
print(classification_report(y_test, yy_test, target_names=target_names))

                 precision    recall  f1-score   support

Below threshold       0.71      0.81      0.76        31
Above threshold       0.76      0.66      0.70        29

       accuracy                           0.73        60
      macro avg       0.74      0.73      0.73        60
   weighted avg       0.74      0.73      0.73        60



Classification rate of 77%, considered as good accuracy.

F1-score: harmonic mean of precision and recall.

Precision: Precision is about being precise, i.e., how accurate your model is. In other words, you can say, when a model makes a prediction, how often it is correct. In this prediction case, when your Logistic Regression model predicted pregnant women are going to have a c-section above the.....

Recall: If there are patients who have a c-section in the test set and your Logistic Regression model can identify it 74% of the time.